### PR TITLE
update examples to cuda native architecture

### LINKED ISSUE #135: Couette flow analytic solution and boundary
- add a couette flow example
- add an analytic_solution
- update boundary condition for top and bottom plates (currently too low at `< 1e-6`)

### LINKED ISSUE #162: Better Examples
Currently, the examples section is tedious to work through. We should enumerate the files and move more advanced examples to a subfolder.

### LINKED ISSUE #184: Reporter likely to fail
Simulation.__call__() calls Reporter(self.flow) but should call Reporter(self), as Reporter expects the simulation class rather than flow

https://github.com/rawbby/lettuce/blob/eefbe699a6bd6a32aa73232defb3ed9bc3a4021e/lettuce/_simulation.py#L57
https://github.com/rawbby/lettuce/blob/eefbe699a6bd6a32aa73232defb3ed9bc3a4021e/lettuce/_simulation.py#L195

### LINKED ISSUE #198: Update examples to cuda-native architecture

### LINKED ISSUE #199: move boundaries back to lt.Flow
lt.Object is a prime example that the boundaries list is better off in lt.Flow, where flow._mask defines the boundary position. This would be possible in flow, but moves the boundary definition into the main file, which bloats it unnecessarily.

### LINKED ISSUE #201: boundaries are only executed with no_collision_mask
make_no_streaming_mask is only used by Simulation to overlap no-streaming areas. It does not affect the no_collision_mask storing the BC identifiers!

### LINKED ISSUE #209: Pass flow to force
forces would often benefit from being passed the lt.Flow instead of only f.

### PR DESCRIPTION
**Pull request by PhiSpel**
## Description
- updated examples to cuda-native version
- structured examples for a nice overview
- added some more simple flow examples
- moved BC definitions into lt.Flow to handle only flow._mask in main
- made reporters use simulation in call
- made observables use simulation and optional observable_parameters in call -> no information required in init

closes #209 
closes #201 
closes #199 
closes #198
closes #184 
closes #162 
closes #135

## TODO

- [x] Fix AntiBounceBackOutlet (add collision to flow so that AntiBounceBackOutlet can use it to stabilize)
- [x] fix EquilibriumOutletP (einsum issue #111 )
- [x] decide on stable outlet boundary for lt.Obstacle
- [x] adapt tests to define boundary in flow
- [x] include boundaries property in all flows
- [x] adapt all lt.Reporter subclasses to use simulation instead of flow
- [x] adapt all lt.Observable subclasses to use simulation instead of f, and only in __call__, not in __init__
- [x] check all tests still run well
  - [x] test_equilibrium_boundary_pu_algorithm throws assertion error

## Checklist

 - [x] This pull request is associated to an issue
 - [x] This PR contains a description
 - [x] Add someone else as reviewer and wait for approval before merging.

### CODE REVIEW

**Comment by PhiSpel** [comment]
Convergence issue is most likely due to

- [ ] ErrorReporter,
- [ ] BGKCollision, or
- [ ] TGVFlow.

Will check this when I get back to it.

**Comment by PhiSpel** [comment]
@McBs, have fun! :)

**Comment by PhiSpel** [comment]
_edit_: making TorchStencil private appears to be not so simple, so I reveretd it and will deal with it separately

File: lettuce/ext/_flows/taylorgreen.py [hunk 1]
```diff
@@ -1,85 +1,119 @@
...
+                                   device=self.context.device,
+                                   dtype=self.context.dtype)
+                    for n in range(self.d))
+        return torch.meshgrid(*xyz, indexing='ij')
+
+    def initial_pu(self) -> (torch.Tensor, torch.Tensor):
         return self.analytic_solution(t=0)
 
-    def analytic_solution(self, t=0):
+    def analytic_solution(self, t: float) -> (torch.Tensor, torch.Tensor):
```
Comment (McBs) [inline lettuce/ext/_flows/taylorgreen.py:65]: Add a warning when `stencil.d == 3` and the analytic solution is called since its only true for 2D

Comment (McBs) [reply lettuce/ext/_flows/taylorgreen.py:65]: ah, at least, when the time is `t>0 and stencil.d ==3`

File: old_tests/test_reporters.py [hunk 2]
```diff
@@ -89,10 +102,10 @@ def test_HDF5Reporter(tmpdir):
     hdf5_reporter = HDF5Reporter(
         flow=flow,
         collision=collision,
-        interval=1,
-        filebase= tmpdir / "output")
+        interval=step,
+        filebase=tmpdir / "output")
     simulation.reporters.append(hdf5_reporter)
-    simulation.step(3)
+    simulation.step(step)
```
Comment (McBs) [inline old_tests/test_reporters.py:108]: `simulation(step)` since the step method is deprecated

File: old_tests/test_simulation.py [hunk 3]
```diff
@@ -77,15 +84,19 @@ def test_initialize_fneq(Case, dtype_device):
...
     if Case is TaylorGreenVortex2D:
-        error_reporter_neq = ErrorReporter(lattice, flow, interval=1, out=None)
-        error_reporter_eq = ErrorReporter(lattice, flow, interval=1, out=None)
-        simulation_eq = Simulation(flow, lattice, collision, streaming)
-        simulation_neq.reporters.append(error_reporter_neq)
-        simulation_eq.reporters.append(error_reporter_eq)
+        error_reporter_neq = ErrorReporter(flow.analytic_solution, interval=1,
+                                           out=None)
+        error_reporter_eq = ErrorReporter(flow.analytic_solution,
+                                          interval=1, out=None)
```
Comment (McBs) [inline old_tests/test_simulation.py:90]: line breaks differ between 87 and 89.

**Review by McBs — CHANGES_REQUESTED** [review]
see comments and PRs

File: lettuce/ext/_flows/taylorgreen.py (hunk 1 above)
Comment (PhiSpel) [reply lettuce/ext/_flows/taylorgreen.py:65]: see [0297d37](https://github.com/lettucecfd/lettuce/pull/200/commits/0297d37fb2785f9ad8ac347fe3e475a9e89927b1)

File: old_tests/test_reporters.py (hunk 2 above)
Comment (PhiSpel) [reply old_tests/test_reporters.py:108]: this will be adapted as part of #213

File: old_tests/test_simulation.py (hunk 3 above)
Comment (PhiSpel) [reply old_tests/test_simulation.py:90]: this will also be adapted in #213

**Review by McBs — APPROVED** [review]
lgtm. Thanks!

In future PRs, let's aim to address smaller issues. This will help us identify bugs more quickly and ensure a smoother review process

**Pull request merged**